// cacache/lib/util/tmp.js
'use strict'

const BB = require('cacache/lib/util/node_modules/cacache/lib/util/node_modules/bluebird')

const figgyPudding = require('cacache/lib/util/node_modules/cacache/lib/util/node_modules/figgy-pudding')
const fixOwner = require('cacache/lib/util/node_modules/cacache/lib/util/fix-owner')
const path = require('path')
const rimraf = BB.promisify(require('cacache/lib/util/node_modules/cacache/lib/util/node_modules/rimraf'))
const uniqueFilename = require('cacache/lib/util/node_modules/cacache/lib/util/node_modules/unique-filename')

const TmpOpts = figgyPudding({
  tmpPrefix: {}
})

module.exports.mkdir = mktmpdir
function mktmpdir (cache, opts) {
  opts = TmpOpts(opts)
  const tmpTarget = uniqueFilename(path.join(cache, 'tmp'), opts.tmpPrefix)
  return fixOwner.mkdirfix(cache, tmpTarget).then(() => {
    return tmpTarget
  })
}

module.exports.withTmp = withTmp
function withTmp (cache, opts, cb) {
  if (!cb) {
    cb = opts
    opts = null
  }
  opts = TmpOpts(opts)
  return BB.using(mktmpdir(cache, opts).disposer(rimraf), cb)
}

module.exports.fix = fixtmpdir
function fixtmpdir (cache) {
  return fixOwner(cache, path.join(cache, 'tmp'))
}
